### DISCUSSION TITLE
why not the jupytext format?

### DISCUSSION THREAD

**Discussion by majidaldo** [marimo-team/marimo#2148] [Ideas]
https://github.com/marimo-team/marimo/issues/2124

From what i can tell, marimo is parsing code for variables to assign them to input and output. So why not just stick with ~jupytext? If io is going to be parsed anyways, why need a "format" for marimo?

**Comment by akshayka** [comment]
You mean the file format, with @app.cell decorators? It's to make notebook files lazy so that they can be imported without executing the entire notebook, while still making the code accessible for re-use (ie not locked away in an if name ==  __main__ guard). We have limited support for notebook reusability/composition (see the App and Cell sections in our API docs: https://docs.marimo.io/api/index.html), with more support planned.

Learn more here: https://github.com/marimo-team/meps/blob/main/mep-0001.md

**Reply by majidaldo** [reply]
[Jupyter grade](https://github.com/marimo-team/meps/blob/main/mep-0001.md#final-verdict ) = F  😆

**Reply by majidaldo** [reply]
still, you can (just) parse the jupytext format and generate `app.cell`s. i'm one of those that want to use my own editor. i can use linting with jupytext but not as well with `@app.cell`s.

**Reply by dmadisetti** [reply]
But you can also just run marimo without any parsing at all. If you want something more like jupytext marimo also has a markdown mode: https://github.com/marimo-team/marimo/blob/main/marimo/_tutorials/markdown_format.md (which I'm realizing isn't explicitly in the docs)

What are the issues you have with linting?

**Reply by majidaldo** [reply]
it's all about typing: the ide just doesn't know about `cell` argument types.

**Reply by akshayka** [reply]
> it's all about typing: the ide just doesn't know about `cell` argument types. 

Yes, I agree that's the biggest drawback of our file format. I would love if we had an editor agnostic (LSP?) plugin that inferred types for the cell arguments. Or, we can try to serialize type hints in function args. This gets messy with imported modules.

**Reply by dmadisetti** [reply]
We could record the type on tree traversal if provided. E.g.

```python
a : int = 0
b =  None
```

```python
untyped_response = a if b else 1
```

```python
typed_response: str = a if b else 1
```

gives

```python
@app.cell
def _(): # require all hints for typing
    a : int = 0
    b =  None
    return a, b
```

```python
@app.cell
def _(a: int, b):
    untyped_response = a if b else 1
    return (untyped_response,)
```

```python
@app.cell
def _(a: int, b) -> tuple[str]:
    typed_response: str = a if b else 1  # Don't bother to enforce poor typing, defer to linter, just capture what's put
    return (typed_response,)
```

---

I don't think resolving implicit typing is worth it

**Reply by akshayka** [reply]
Using user types was my first thought. But we already have "implicit" types through jedi/lsp, no?

The only thing that's troublesome to type is modules. Eg, if np is an arg to a cell. You'd want completion.

**Reply by dmadisetti** [reply]
I think a reasonable compromise is the setup cell for types and modules

**Reply by akshayka** [reply]
Yea i think that could be fine. With setup cell for modules, plus ideally inferred types (in a way that doesn't make saving slow), we'd have a pretty great editing experience!
